OpenMPI: improve PMix handling

sethrj opened the pull request:
Currently OpenMPI 4+ requires PMIx and provides a vendored copy (version 3.2.3 for OpenMPI 4.1.2) if not found. This change uses Spack's PMIx to provide a dependency and updates the available PMIx versions, as well as removing an artificial constraint on the libevent used by pmix. Finally, since the `pmix` variant is only useful for OpenMPI 2 and 3, I added a `when` clause for it.

(The reason I'm concerned about pmix is our CI server has been seeing hangs in simple single-process global barriers, and occasional crashes coming from pmix error checking code. I'm hoping that a newer version of PMIx might fix these.)

sethrj committed:
openmpi: always require pmix for 4:

`~pmix` is not applicable to version 4+, since it always builds a vendored
copy of pmix (currently 3.2.3).

sethrj committed:
pmix: relax version requirements

When the version range was specified, newer versions didn't exist.
Also use normalized spack versions rather than artificial .9.9 /.0.0.

sethrj committed:
openmpi: restrict pmix versions

pmix option isn't available for OpenMPI@1, and according to
https://github.com/open-mpi/ompi/issues/7988 , OpenMPI 4.0.1 will not
build with pmix@3.1.5.

sethrj committed:
pmix: add newer versions

rhc54 commented:
Errr...we haven't heard of any problems with PMIx v3 or above running with OMPI. Can you perhaps clarify and/or provide some details?

sethrj commented:
Sure, not at my computer right now so I'll give you a sample later. The whole stack is built with spack and I can't say for certain whether the error is in PMIx (though some of the crashes have backtraces pointing there). But in general in spack it's better to explicitly depend on packages than vendor them.

rhc54 commented:
No argument there - we expect that approach from packagers. My only interest was to discover what problems you are seeing relative to PMIx with OMPI so we can fix them.

For clarity, OMPI v4 does not _require_ PMIx - you can still use PMI-1/2 with that series. It is OMPI v5 and above that only supports PMIx.

sethrj commented:
Ah ok, thanks for the explanation. The current behavior seems to be to install pmix even if `~pmix`, rather than fall back to pmi.

sethrj commented:
@rhc54 Here are two example failures; this happened in a simple test launched with `mpiexec "-n" "1" "--bind-to" "none"`, at the end of the harness (I think either at `MPI_Barrier` or `MPI_Reduce_scatter`) but before exiting `main()`.
```
[scalel64-7:31510] *** Process received signal ***
[scalel64-7:31510] Signal: Segmentation fault (11)
[scalel64-7:31510] Signal code: Address not mapped (1)
[scalel64-7:31510] Failing at address: 0x150
[scalel64-7:31510] [ 0] /lib64/libpthread.so.0[0x351da0f7e0]
[scalel64-7:31510] [ 1] /projects/spack/opt/spack/gcc-8.5.0/openmpi/lt6jzov/lib/libopen-pal.so.40(pmix_notify_check_range+0x8)[0x7f077b738ce8]
[scalel64-7:31510] [ 2] /projects/spack/opt/spack/gcc-8.5.0/openmpi/lt6jzov/lib/libopen-pal.so.40(+0x17e030)[0x7f077b73a030]
[scalel64-7:31510] [ 3] /projects/spack/opt/spack/gcc-8.5.0/libevent/5i47rbn/lib/libevent_core-2.1.so.7(+0x1e909)[0x7f077b585909]
[scalel64-7:31510] [ 4] /projects/spack/opt/spack/gcc-8.5.0/libevent/5i47rbn/lib/libevent_core-2.1.so.7(event_base_loop+0x4ff)[0x7f077b5861cf]
[scalel64-7:31510] [ 5] /projects/spack/opt/spack/gcc-8.5.0/openmpi/lt6jzov/lib/libopen-pal.so.40(+0x1be06e)[0x7f077b77a06e]
[scalel64-7:31510] [ 6] /lib64/libpthread.so.0[0x351da07aa1]
[scalel64-7:31510] [ 7] /lib64/libc.so.6(clone+0x6d)[0x351d6e8c4d]
[scalel64-7:31510] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 0 on node scalel64-7 exited on signal 11 (Segmentation fault).
```
and
```
[cloud-dev02:2800498] *** Process received signal ***
[cloud-dev02:2800498] Signal: Segmentation fault (11)
[cloud-dev02:2800498] Signal code: Address not mapped (1)
[cloud-dev02:2800498] Failing at address: 0x7f130000014f
[cloud-dev02:2800498] [ 0] /lib64/libpthread.so.0(+0xf630)[0x7f1216e07630]
[cloud-dev02:2800498] [ 1] /projects/spack/opt/spack/gcc-8.5.0/openmpi/tba7x4q/lib/libopen-pal.so.40(pmix_notify_check_range+0x8)[0x7f120cced3a8]
[cloud-dev02:2800498] [ 2] /projects/spack/opt/spack/gcc-8.5.0/openmpi/tba7x4q/lib/libopen-pal.so.40(+0x1796f0)[0x7f120ccee6f0]
[cloud-dev02:2800498] [ 3] /projects/spack/opt/spack/gcc-8.5.0/libevent/3qfrfg3/lib/libevent_core-2.1.so.7(+0x1dc29)[0x7f120c550c29]
[cloud-dev02:2800498] [ 4] /projects/spack/opt/spack/gcc-8.5.0/libevent/3qfrfg3/lib/libevent_core-2.1.so.7(event_base_loop+0x4ff)[0x7f120c5514ef]
[cloud-dev02:2800498] [ 5] /projects/spack/opt/spack/gcc-8.5.0/openmpi/tba7x4q/lib/libopen-pal.so.40(+0x1b972e)[0x7f120cd2e72e]
[cloud-dev02:2800498] [ 6] /lib64/libpthread.so.0(+0x7ea5)[0x7f1216dffea5]
[cloud-dev02:2800498] [ 7] /lib64/libc.so.6(clone+0x6d)[0x7f12154aab0d]
[cloud-dev02:2800498] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 0 on node cloud-dev02 exited on signal 11 (Segmentation fault).
```

These are on two different systems, with two different OS (RHEL6 and CentOS7), two different processors but on generic (x86_64) builds:
```
tba7x4q openmpi@4.1.2~atomics~cuda~cxx~cxx_exceptions+gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~pmi~pmix+romio~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=none schedulers=none
lik7l7z     hwloc@2.7.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared
4x56q3s         libpciaccess@0.16
5k5aenm         libxml2@2.9.12~python
a5ut2ia             libiconv@1.16 libs=shared,static
6gols7g             xz@5.2.5~pic libs=shared,static
draacp6             zlib@1.2.11+optimize+pic+shared
ppry2fz         ncurses@6.2~symlinks+termlib abi=none
3qfrfg3     libevent@2.1.12+openssl
qqgwhbn         openssl@1.1.1m~docs certs=system
ira3u7c     numactl@2.0.14 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94,62fc8a8bf7665a60e8f4c93ebbd535647cebf74198f7afafec4c085a8825c006,ff37630df599cfabf0740518b91ec8daaf18e8f288b19adaae5364dc1f6b2296
e7wka2b     openssh@7.4p1
```

These tests are being linked against googletest and a few other scientific libraries, but it's doing so using their built-in CMake scripts and inside the same spack environment -- so the toolchain is consistent. This was also happening sporadically (but I think less frequently) in a separate `openmpi@4.0.5` spack build. The failure rate (by crash or hang) is about 1 in 1000 if I had to guess. I've run some of these tests with `ctest --repeat-until-fail` but with no luck, so it might also have to do with a high server load.

If you'd like me to open an issue on OpenMPI to discuss this I'm very happy to --  but I've been pulling my hair out trying to narrow down potential causes, and I thought that using a more predictable (and possibly newer and more fixed) version of pmix could help.

Thanks!

rhc54 commented:
Yeah, I would definitely encourage you to open those against OMPI. I'm not sure why OMPI would be calling `PMIx_Notify` if all it is doing is an `MPI_Barrier`. Given that they run nightly regressions across thousands of tests, it seems a little suspicious that you'd see a failure on something that trivial.

hppritcha commented:
The proposed changes to the pmix and openmpi packages look reasonable.  I"m not sure this PR is the right place to try and find a resolution to the failures being observed.  As a started I'd suggest building OPen MPI and the external PMIx with debug enabled and see if core files might provide more information.

hppritcha commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -319,14 +319,16 @@ class Openmpi(AutotoolsPackage):
 
+    # PMIx is required for OpenMPI 4; omitting will use an internal copy
+    # Vendored version: depends_on('pmix@3.2.3', when='@4.1.2')
+    depends_on('pmix@3.2:', when='@4:')
```
this will get more complicated soon as an eminent release of Open MPI uses features only available starting in PMIx 4.

hppritcha replied:
actually this depends_on here should be conditional on +pmix being in the spec as well when at 4 or older.

sethrj commented:
> For clarity, OMPI v4 does not _require_ PMIx - you can still use PMI-1/2 with that series. It is OMPI v5 and above that only supports PMIx.

Should I change this from making `pmix` a requirement for @4 to requiring it when `~pmi`?

>  it seems a little suspicious that you'd see a failure on something that trivial

Yes, the failure has the smell of UB from a code bug or linking error on our end, but I haven't found anything suspicious in what we're doing: and I think it only started when our CI switched to MPI 4. I'll open up an issue to see if anyone's seen anything similar.

> I'm not sure this PR is the right place to try and find a resolution to the failures being observed. As a started I'd suggest building OpenMPI and the external PMIx with debug enabled and see if core files might provide more information.

Sure, I think it's unlikely that this will help -- but I thought the builtin pmix smelled a little funny, and this could help with debugging.

rhc54 commented:
> Sure, I think it's unlikely that this will help -- but I thought the builtin pmix smelled a little funny, and this could help with debugging.

It would also allow you to use an updated PMIx beyond what is in the OMPI v4 series - might help to see if the more current PMIx releases resolve the problem.

hppritcha reviewed (changes requested):
the proposed depends_on at line 324 should also be dependent on +pmix for @2:4

sethrj commented:
> the proposed depends_on at line 324 should also be dependent on +pmix for @2:4

There's already a more general pmix dependency above -- I'll conslidate it (and libevent) with this to be less confusing.

sethrj committed:
OpenMPI: re-express conflicts/configure logic as conditional variants

This relies partly on `self.enable_or_disable` and its ilk to emit an
empty list when the variant isn't applicable.

sethrj commented:
@hppritcha I took the opportunity to refactor much of the other version dependencies (conflics, depends, variant changes) based on the relatively new `when=` option for variants. This should simplify the package going forward.

rhc54 commented:
Something to potentially keep in mind: PMIx ensures compatibility as you go upward in versions. In other words, if OMPI vX works with PMIx vY, then it will also work with PMIx vY+. So if you find that a particular version of OMPI is displaying problems with its internal version level, you could check to see if it works better with a higher version - and then set that higher version as the minimum dependency.

Rationale being that PMIx may have moved beyond its support limits on the older versions. For instance, we no longer support the PMIx v2 series at all, which is what OMPI v3 contains. Once PMIx v5 releases (probably this summer), we'll be dropping support for PMIx v3, which is what OMPI v4 contains.

And as was noted, OMPI v5 is probably going to require PMIx v4.2+ for its anticipated feature set.

sethrj commented:
For what it's worth, our CI tests are still seeing sporadic crashes with `openmpi@4.1.2 ^pmix@4.1.2 ^libevent@2.1.12`:
```
[cloud-dev01:2917909] *** Process received signal ***
[cloud-dev01:2917909] Signal: Segmentation fault (11)
[cloud-dev01:2917909] Signal code: Address not mapped (1)
[cloud-dev01:2917909] Failing at address: 0x7fab0000014f
[cloud-dev01:2917909] [ 0] /lib64/libpthread.so.0(+0xf630)[0x7faab7908630]
[cloud-dev01:2917909] [ 1] /projects/spack/opt/spack/gcc-8.5.0.static/pmix/7p6hkue/lib/libpmix.so.2(pmix_notify_check_range+0x8)[0x7faaabf01e08]
[cloud-dev01:2917909] [ 2] /projects/spack/opt/spack/gcc-8.5.0.static/pmix/7p6hkue/lib/libpmix.so.2(+0x402c8)[0x7faaabf032c8]
[cloud-dev01:2917909] [ 3] /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_core-2.1.so.7(+0x1dc49)[0x7faaab687c49]
[cloud-dev01:2917909] [ 4] /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_core-2.1.so.7(event_base_loop+0x4ff)[0x7faaab68850f]
[cloud-dev01:2917909] [ 5] /projects/spack/opt/spack/gcc-8.5.0.static/pmix/7p6hkue/lib/libpmix.so.2(+0xdbf2e)[0x7faaabf9ef2e]
[cloud-dev01:2917909] [ 6] /lib64/libpthread.so.0(+0x7ea5)[0x7faab7900ea5]
[cloud-dev01:2917909] [ 7] /lib64/libc.so.6(clone+0x6d)[0x7faab656fb0d]
[cloud-dev01:2917909] *** End of error message ***
```

aweits commented:
> For what it's worth, our CI tests are still seeing sporadic crashes with `openmpi@4.1.2 ^pmix@4.1.2 ^libevent@2.1.12`:

I've run into something like this before - are you sure you're not inadvertently pulling in different (multiple) libevents at runtime? (what does ldd look like on the binary?)

sethrj commented:
> I've run into something like this before - are you sure you're not inadvertently pulling in different (multiple) libevents at runtime? (what does ldd look like on the binary?)

UB from mixed-up linking was my first thought, so I've been very careful to check. Everything in this toolchain is consistent, using only files from a single spack concretized-together environment.
```
ldd tstJoin | grep event
	libevent_core-2.1.so.7 => /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_core-2.1.so.7 (0x00007f429d9a1000)
	libevent_pthreads-2.1.so.7 => /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_pthreads-2.1.so.7 (0x00007f429d79e000)
```

ADDENDUM: and the `image list` from the core dump (https://github.com/openpmix/openpmix/issues/2508) shows nothing out of place either. (The GCC 8.5 version is built using the system GCC 4.8, which is why you see the occasional `gcc-4.8.5` in the list.)
<details><pre>
[  0] CF73C083 0x0000000000400000 /home/s3j/mpi-failure/tstJoin
[  1] 3601C3B6-C6AC-D44D-4554-A045F4D093CB-6D343189 0x00007ffea6df2000 [vdso] (0x00007ffea6df2000)
[  2] EF841190 0x00007faab88a2000 /gitlab/iyguzBfw/0/rnsd/scale/build/src/nemesis/libNemesisGtest.so
[  3] 37E9BC46 0x00007faab826f000 /gitlab/iyguzBfw/0/rnsd/scale/build/lib/libnemesis.so
[  4] 0C9B91A8 0x00007faab7d43000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libmpi.so.40
[  5] 672394EE 0x00007faab7b15000 /projects/spack/opt/spack/gcc-4.8.5/gcc/y245fnv/lib64/libgomp.so.1
[  6] E10CC8F2-B932-FC3D-AEDA-22F8DAC5EBB9-69524E5B 0x00007faab78f9000 /lib64/libpthread.so.0
[  7] F163F3C3 0x00007faab7339000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libhdf5.so.103
[  8] 667089FC 0x00007faab70dd000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib64/libgtest.so
[  9] 4635B531 0x00007faab6d59000 /projects/spack/opt/spack/gcc-4.8.5/gcc/y245fnv/lib64/libstdc++.so.6
[ 10] 7615604E-AF4A-068D-FAE5-085444D15C0D-EE93DFBD 0x00007faab6a57000 /lib64/libm.so.6
[ 11] FB9706EC 0x00007faab683f000 /projects/spack/opt/spack/gcc-4.8.5/gcc/y245fnv/lib64/libgcc_s.so.1
[ 12] 8DA9707D-518E-8817-9B1B-2903CD846C82-102D84C1 0x00007faab6471000 /lib64/libc.so.6
[ 13] 2B5C6B16 0x00007faab6214000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libkokkoscore.so.13
[ 14] 7F2E9CB0-769D-7E57-BD66-9B485A74B537-B63A57C4 0x00007faab6010000 /lib64/libdl.so.2
[ 15] 19D57FEB 0x00007faab5db5000 /projects/spack/opt/spack/gcc-8.5.0.static/hwloc/kgmp3uz/lib/libhwloc.so.15
[ 16] C4BB82FD 0x00007faab5b94000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libhdf5_hl.so.100
[ 17] 65674C7D 0x00007faab598d000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libanasazitpetra.so.13
[ 18] 9454E9B7 0x00007faab5359000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libbelostpetra.so.13
[ 19] 767DBE29 0x00007faab4ddb000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libbelosepetra.so.13
[ 20] 78AD3C9A 0x00007faab49ef000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libbelos.so.13
[ 21] BBFFD48E 0x00007faab4774000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libaztecoo.so.13
[ 22] 7BB9D3DD 0x00007faab4553000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libthyratpetra.so.13
[ 23] D84B5A40 0x00007faab42d8000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libthyraepetraext.so.13
[ 24] 1BF77A10 0x00007faab405e000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libthyraepetra.so.13
[ 25] CE3E0203 0x00007faab3b8c000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libthyracore.so.13
[ 26] 8B61B2C3 0x00007faab3956000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/librtop.so.13
[ 27] 8643F735 0x00007faab3655000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libamesos2.so.13
[ 28] EE3CD586 0x00007faab332c000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetraext.so.13
[ 29] 62D88250 0x00007faab3101000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetrainout.so.13
[ 30] 76035015 0x00007faab2613000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetra.so.13
[ 31] 0313D4DA 0x00007faab238c000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libkokkostsqr.so.13
[ 32] 9C84A29B 0x00007faab1c02000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libkokkoskernels.so.13
[ 33] 2234E320 0x00007faab19ff000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetraclassiclinalg.so.13
[ 34] C1E95118 0x00007faab17fc000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetraclassicnodeapi.so.13
[ 35] C0CC1AE0 0x00007faab15f9000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtpetraclassic.so.13
[ 36] 8694ECCD 0x00007faab12aa000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libepetraext.so.13
[ 37] A5777737 0x00007faab1060000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtriutils.so.13
[ 38] CA0E0D64 0x00007faab0cfe000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libepetra.so.13
[ 39] 3684D947 0x00007faab0afb000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchoskokkoscomm.so.13
[ 40] 32902B14 0x00007faab08f8000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchoskokkoscompat.so.13
[ 41] C94AD6CC 0x00007faab06f1000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchosremainder.so.13
[ 42] 6E0F3AB2 0x00007faab04c2000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libtrilinosss.so.13
[ 43] BD081223 0x00007faab02bf000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libkokkosalgorithms.so.13
[ 44] 0D32B292 0x00007faab00bc000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libkokkoscontainers.so.13
[ 45] D98FAD38 0x00007faaafea4000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchosnumerics.so.13
[ 46] 0C522C88 0x00007faaafbac000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchoscomm.so.13
[ 47] E23649F8 0x00007faaaf566000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchosparameterlist.so.13
[ 48] 288449F6 0x00007faaaf2f5000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchosparser.so.13
[ 49] DE5BC9E5 0x00007faaaf09a000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libteuchoscore.so.13
[ 50] B620A684 0x00007faaac95e000 /projects/spack/var/spack/environments/detribits/.spack-env/view/lib/libopenblas.so.0
[ 51] 62C44997-4331-341B-B08D-CCE3859560A2-2AF1E172 0x00007faab8aac000 /lib64/ld-linux-x86-64.so.2
[ 52] 23948BB2 0x00007faaac640000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libopen-rte.so.40
[ 53] 38392AB7 0x00007faaac32d000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libopen-pal.so.40
[ 54] A7AA77F5 0x00007faaabec3000 /projects/spack/opt/spack/gcc-8.5.0.static/pmix/7p6hkue/lib/libpmix.so.2
[ 55] 3E44DF70-5594-2478-D052-E40FDD1F5B78-62B152B0 0x00007faaabcbb000 /lib64/librt.so.1
[ 56] FF2196BD-22A8-4430-54C8-3031E0E76EB0-1BA1219C 0x00007faaabab8000 /lib64/libutil.so.1
[ 57] 17AA147D 0x00007faaab8a1000 /projects/spack/opt/spack/gcc-8.5.0.static/zlib/iyy6cky/lib/libz.so.1
[ 58] 1A07CF95 0x00007faaab66a000 /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_core-2.1.so.7
[ 59] 5D62F3A0 0x00007faaab467000 /projects/spack/opt/spack/gcc-8.5.0.static/libevent/iebojck/lib/libevent_pthreads-2.1.so.7
[ 60] 56A2DF3C 0x00007faaab25e000 /projects/spack/opt/spack/gcc-8.5.0.static/libpciaccess/2lkmzcq/lib/libpciaccess.so.0
[ 61] E9AB1AA7 0x00007faaaaefb000 /projects/spack/opt/spack/gcc-8.5.0.static/libxml2/lkze2cg/lib/libxml2.so.2
[ 62] 5F14985B 0x00007faaaacbd000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libmpi_usempif08.so.40
[ 63] BBA55A65 0x00007faaaaaaf000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libmpi_usempi_ignore_tkr.so.40
[ 64] 05BE8940 0x00007faaaa84b000 /projects/spack/opt/spack/gcc-8.5.0.static/openmpi/5az5n35/lib/libmpi_mpifh.so.40
[ 65] 8B2B683E 0x00007faaaa3d9000 /projects/spack/opt/spack/gcc-4.8.5/gcc/y245fnv/lib64/libgfortran.so.5
[ 66] 2326A8DE 0x00007faaaa199000 /projects/spack/opt/spack/gcc-4.8.5/gcc/y245fnv/lib64/libquadmath.so.0
[ 67] 4C77C427 0x00007faaa9f73000 /projects/spack/opt/spack/gcc-8.5.0.static/xz/6s2r2ue/lib/liblzma.so.5
[ 68] 3216FFD3 0x00007faaa9c77000 /projects/spack/opt/spack/gcc-8.5.0.static/libiconv/az6pdin/lib/libiconv.so.2
</pre></details>

sethrj commented:
@aweits still digging around: could https://github.com/open-mpi/ompi/pull/8792 be the root cause? With spack we're definitely using external *everything*...

sethrj commented:
@hppritcha @haampie could you review this? The conversation got a bit off track and it's clear the changes don't help with the issue I'm seeing, but changing `pmix` from a vendored to an explicit dependency is good, and I think the changes to the recipe substantially simplify the package package

aweits commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -354,11 +348,21 @@ class Openmpi(AutotoolsPackage):
 
     conflicts('schedulers=slurm ~pmi', when='@1.5.4:',
               msg='+pmi is required for openmpi(>=1.5.5) to work with SLURM.')
```
I think this should be:

```suggestion
    conflicts('schedulers=slurm ~pmi', when='@1.5.4:2',
              msg='+pmi is required for openmpi to work with SLURM.')
    conflicts('schedulers=slurm ~pmi ~pmix', when='@3:',
              msg='+pmi or +pmix is required for openmpi to work with SLURM.')
```

hppritcha replied:
I agree this suggested change is more accurate.

sethrj replied:
It seems that this change is causing problems for @haampie 's slurm-based build of openmpi:
```
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
```

His previous build had `+pmi` because of the original conflict as written. It might be that newer OpenMPI versions still require `+pmi` for older versions of SLURM: @aweits can you provide clarification on your suggestion?

rhc54 replied:
Unfortunately, that will not be true starting with OpenMPI v5.0 as _only_ PMIx will be supported - OMPI will no longer support legacy PMI versions. For older versions of OMPI, you can still use PMI-1 and PMI-2 with legacy PMI so long as you explicitly configure OMPI with those options. The explicit configure is required due to the GPL nature of those legacy PMI libraries - it protects OMPI from the GPL requirements (passing it along to whomever distributes the result).

This is something you may need to consider. If Spack is distributing binary rpms that have OMPI linked to Slurm PMI-1 or PMI-2, then you (Spack) are responsible for the associated GPL conditions. You should probably check with an appropriate legal person as to how the ramifications might impact you.

aweits replied:
@sethrj did that answer the question? On our side we build slurm as an external w/ pmix, and desire to match that on the openmpi side (i.e. --without-pmi --with-pmix=<path to pmix>).

sethrj replied:
Your last update answers it. In your situation the suggestion is correct but in general it's not. The conflicts should probably be more like:
```python
    conflicts('~pmi', when='^slurm+pmi')
    conflicts('~pmix', when='@3: ^slurm+pmix')
```
The problem with this representation is that I think most users have SLURM installed as an "external" package on an existing server, so Spack has no knowledge of the pmi/pmix variants in use by slurm...

sethrj committed:
Update var/spack/repos/builtin/packages/openmpi/package.py

Co-authored-by: Andrew W Elble <aweits@rit.edu>

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into openmpi-pmix

sethrj commented:
@hppritcha @haampie @aweits Could one of you please review this PR or close it with a comment? This package gets enough updates that a long-lived branch means frequent and difficult conflict resolution. Thanks!

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -226,21 +226,24 @@ class Openmpi(AutotoolsPackage, CudaPackage):
-    variant('pmi', default=False, description='Enable PMI support')
-    variant('pmix', default=False, description='Enable PMIx support')
-    variant('wrapper-rpath', default=True,
+    variant('cuda', default=False, when='@1.7:', description='Enable CUDA support')
```
I believe there's a bug in spack when overriding unconditional variants from the base class with a conditional variant in the derived class; it seems it's effectively unconditional. Can you turn it into a conflict?

sethrj replied:
Thanks for the info, I've updated accordingly.

sethrj committed:
openmpi: cuda now conflict not conditional

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -702,33 +693,36 @@ def configure_args(self):
+
         # Hwloc support
-        if '~internal-hwloc' in spec and spec.satisfies('@1.5.2:'):
+        if 'hwloc' in spec:
```
Maybe explicit `^hwloc` to make it more clear it applies to a dependency?

sethrj replied:
I wonder if the `internal-hwloc` variant should be deleted... seems like it's asking for a conflict, and as a bonus the variant name uses the ambiguous `-` rather than `_`.

haampie reviewed (dismissed):
It's nice some variants are now conditional. I did not check all versions for correctness, but I trust @sethrj's research

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -318,27 +320,24 @@ class Openmpi(AutotoolsPackage, CudaPackage):
+    # CUDA support was added in 1.7, and since the variant is part of the
+    # parent package we must express as a conflict rather than a conditional
+    # variant.
+    conflicts('+cuda', when=':@1.6')
```
```suggestion
    conflicts('+cuda', when='@:1.6')
```

haampie committed:
Typo

sethrj committed:
Change spec usage

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into openmpi-pmix

sethrj commented:
@haampie I had to merge develop into the branch again in order to fix the `spack spec -I` error so I could test the build. I think this should be the last time...

aweits reviewed (approved):
builds/runs/tested 4.0.6 and 4.1.3
